### ISSUE TITLE
List of necessary material for documentation

### ISSUE THREAD

**Issue by leios** [JuliaGPU/CUDA.jl#5]
I would like to help a bit with documentation and would like to create a list of things we need to discuss to "complete" the documentation. Issue #1 brought up installation instructions, but I thought a more defined list would be worthwhile.

In particular, I feel we are missing:

1. Shared memory -- Here, we seem to be missing use-cases for both static and dynamic shared memory. For this, I would propose a worked example of a [simple (single-GPU) transpose](https://devblogs.nvidia.com/efficient-matrix-transpose-cuda-cc/)
2. Using different streams -- Maybe recreating [this](https://devblogs.nvidia.com/gpu-pro-tip-cuda-7-streams-simplify-concurrency/) example? Actually, the only way I know how to initialize different streams is via CUDAdrv. Is there a better way to do this?
3. Using the CuFFT library -- This is especially useful because if you do not create your own plan and are doing any form of timestepping, the garbage collector cannot always keep up and you can quickly run out of memory.
4. Cuda-aware MPI example in the multi-GPU example. I can see this not being useful to most users, though.

As a general question, would it be worth adding a section in the documentation for examples, explicitly?

**Comment by dnabanita7** [comment]
I will add these. Adding examples will surely help to make the users/developers understand and they can tweak around finding errors. So, I think adding it will surely do good.

**Comment by leios** [comment]
Hey Naba, I also have experience with this and don't mind helping out. I just wanted to see if there were further examples people wanted.

**Comment by dnabanita7** [comment]
Hey @leios ! Oh sure! You can add them. I should complete the cuda installation at the first place.

**Comment by maleadt** [comment]
> As a general question, would it be worth adding a section in the documentation for examples, explicitly?

Sure! There's a tutorials section, so examples would make sense too. One thought for examples would be to use Literate so that the files are regular, executable Julia scripts. But nothing is set in stone, so go ahead and we can always evaluate afterwards.

> Using different streams -- Maybe recreating [this](https://devblogs.nvidia.com/gpu-pro-tip-cuda-7-streams-simplify-concurrency/) example? Actually, the only way I know how to initialize different streams is via CUDAdrv. Is there a better way to do this?

Streams are tricky; we have all the low-level building blocks but haven't threaded it through so that high-level functionality (think CUBLAS or CUFFT calls) use them. So this would require some work first.

As a general comment though, I've always held back on adding documentation on things that are well-documented by CUDA itself. That approach has not shown to work well though,  because lots of users start programming with GPUs in Julia without having CUDA C experience. Still, I think it would hurt maintainability if we copy too much of the upstream documentation here.

**Comment by leios** [comment]
Right, the main thing is that certain keywords are sometimes a bit hard to find in Julia. I don't think redoing an nvidia guide is worthwhile, but having complementary code could be useful.

**Comment by ali-ramadhan** [comment]
> One thought for examples would be to use Literate so that the files are regular, executable Julia scripts.

Another benefit of using Literate.jl is that documentation will fail to build if the examples break (or go out of date).
